github/basic_auth_token.go

```diff
@@ -0,0 +1,199 @@
+package github
+// Provides a TokenSource that uses the basic auth mechanism to get tokens
+//
+//	https://developer.github.com/v3/oauth_authorizations/#create-a-new-authorization
+//
+// Sample Usage:
+//
+//	oa2 := oauth2.Config{ ... }
+//
+//	tc, err := github.NewBasicAuthClient(oa2, "zup", "nuch", "my oauth", []string{"public_repo"})
+//
+//	client := github.NewClient(tc)
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+// Creds provides the authorization data needed to do a Basic Auth call to get a Token
+// and implements the TokenSource interface
+//
+//	Username is the user's Basic Auth username
+//	Password is the corresponding password
+//	PostBodyReader is a reader that will produce the desired body for the Post for the auth call
+//
+// api doc showing the call is here:
+//
+//	https://developer.github.com/v3/oauth_authorizations/#create-a-new-authorization
+//
+type Creds struct {
+	Username       string
+	Password       string
+	PostBodyReader io.Reader
+	Config         *oauth2.Config
+}
+
+// Token set username/password and postbody and do basic auth
+func (ba Creds) Token() (tk *oauth2.Token, err error) {
```

Comment:
Having a `TokenSource` that transparently does the token exchange using Basic Auth is certainly clever, and might be something we want to support.  But if we do, we should first implement all of the [OAuth Authorizations](https://developer.github.com/v3/oauth_authorizations) endpoints as a regular API service like everything else in go-github, and then add a thin wrapper around that that does the necessary calls.